doc/design/model_zoo_dev.md

```diff
@@ -0,0 +1,205 @@
+  sqlflow.train("SELECT ...", MyKerasModel, ... into="my_trained_model")
+  auc, f1 = sqlflow.evaluate("SELECT ...", using="my_trained_model", metric=["AUC", "F1"])
+  unittest.assert(auc > 0.9)
+  unittest.assert(f1 > 0.5)
+```
+
+The above exmaple connects to a MySQL database system.  We can connect
+to other supported database systems as well.
+
+```python
+import sqlflow
+
+def test_MyKerasModel():
+  ## 初始化SQLFlow。 project： maxcompute project；AI引擎默认是PAI
+  sqlflow.init(dbhost="maxcompute://[AK]:[SK]@service-corp.odps.aliyun-inc.com/api?curr_project=alifin_jtest_dev&scheme=http")
+  sqlflow.train("SELECT ...", MyKerasModel, ... into="my_trained_model")
+  auc, f1 = sqlflow.evaluate("SELECT ...", using="my_trained_model", metric=["AUC", "F1"])
+  unittest.assert(auc > 0.9)
+  unittest.assert(f1 > 0.5)
+```
+
+We can run these unit tests by typing the following command.
+
+```bash
+pytest
+```
+
+### Test Using SQLFlow Command-line Tool
+
+Like MySQL has a command-line client `mysql`, we provide `sqlflow`.
+Using this command-line tool, we can run the following command to use
+the model definition.
+
+```bash
+$ sqlflow
+sqlflow > SELECT * from iris
+TO TRAIN MyKerasModel
+LABEL class
+INTO my_iris_model;
+```
+
+## Submit Models
+
+
+### Submit Models
+
+```bash
+$ sqlflow
+sqlflow > ADD MODELDEF def_MyKerasModel.py
+checking model is valid ...[OK]
+building model docker image with version v0.12 ... [OK]
+model myKerasModel.py built as hub.docker.com/username/mykerasmodel:v0.12 on localhost:50051
+Model added successfully，You can run it `TO TRAIN hub.docker.com/username/mykerasmodel:v0.12/MyKerasModel`
+```
+
+### Submit to various SQLFlow servers
+
+1. To a local SQLFlow server.
+
+    ```text
+	$ sqlflow
+	sqlflow@localhost:50051 >
+	```
+
+2. To specified SQLFlow server, for example `192.168.1.1:50051`.
+
+    ```text
+	$ sqlflow -host="192.168.1.1:50051" -dbstr="maxcompute://...."
+	```
+
+3. Using a configuration file.
+
+   ```bash
+   vim ~/.sqlflow/config
+   [sqlflow]
+   host=192.168.1.1:50051
+   dbstr=maxcompute://....
+   user=username
+   password=12345
+   ```
+
+### Model Versioning
+
+Everytime we run `ADD MODELDEF`, the system generates a new version:
+
+```bash
+$ sqlflow
+sqlflow > ADD MODELDEF MyKerasModel.py
+model MyKerasModel.py already exists, add a new version 0.13.
+checking model is valid ...[OK]
+building model docker image with version v0.13 ... [OK]
+model myKerasModel.py built as hub.docker.com/username/mykerasmodel:v0.13 on localhost:50051
+Model added successfully，You can run it `TO TRAIN hub.docker.com/username/mykerasmodel:v0.13/MyKerasModel`
+```
+
+### Manage Model Definitions
+
+```bash
+$ sqlflow
+sqlflow > SHOW MODELDEFS;
+sqlflow > SHARE MODELDEF hub.docker.com/username/mykerasmodel:v0.13 TO xiongmu.wy;
+sqlflow > DROP MODELDEF hub.docker.com/username/mykerasmodel:v0.13;
+```
+
+### Manage Trained Models
+
+```Bash
+$ sqlflow
+sqlflow > PUBLISH my_model;
+sqlflow > SHOW TRAINEDMODELS;
```

Comment:
Should it be `SHOW TRAINEDMODELS` or `SHOW TRAINED MODELS`?

Reply:
We need only `SHOW MODELS`. We should delete design details about pre-trained models in this document

Reply:
I don't see the phrase "pre-trained models" appear in this document. How could we delete it?

Reply:
@shendiaomo we can first put the design here, then consider when to implement it.

And, only `SHOW MODELS` may not be enough since:

1. Users should be able to list model definitions and trained models locally.
1. Users should be able to list model definitions that are publicly available in the model zoo.
1. Users should be able to list model definitions that shared from other users.
1. Users should be able to list trained models that shared from other users.

The SQL statements should be able to distinguish those cases.

Reply:
> I don't see the phrase "pre-trained models" appear in this document. How could we delete it?

I mean `TRAINEDMODELS`.

Reply:
> @shendiaomo we can first put the design here, then consider when to implement it.
> 
> And, only `SHOW MODELS` may not be enough since:
> 
> 1. Users should be able to list model definitions and trained models locally.

Agreed. We only need `show trained models` locally now.

> 2. Users should be able to list model definitions that are publicly available in the model zoo.

Yep.

> 3. Users should be able to list model definitions that shared from other users.

Yep.

> 4. Users should be able to list trained models that shared from other users.

That' what I'm worried about. I'm afraid our users will not have such requirements for a long time (several years).
- For tabular data, sharing a `trained model` is meaningful only if users have exactly the same dataset. Is Kaggle players our targeted users?
- For CV/NLP/Speech data, sharing a `trained model` is rather meaningful. Are we prepared to serve those users?
- For recommendation systems, a STOA model architecture is very useful in various application scenarios, a trained model is not.
- Moreover, most users are satisfied with XGBoost, they don't need a shared model definition, not to mention a trained model.
> 
> The SQL statements should be able to distinguish those cases.